### PR TITLE
All fields

### PR DESCRIPTION
**Pull request by adamf321**

### CODE REVIEW

File: src/All.php [hunk 1]
```diff
@@ -0,0 +1,155 @@
+<?php namespace Leean\Acf;
+
+/**
+ * Class to provide helpers for getting a all ACF fields for an entity.
+ */
+class All
```
Comment (mitogh) [inline src/All.php]: I think this should be the `Acf` class instead of `All`

File: src/All.php [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+	}
+
+	/**
+	 * Apply the default transforms if any exist for this field type.
+	 *
+	 * @param array $field The field.
+	 * @return mixed
+	 */
+	private static function apply_default_transform( $field ) {
+		$class = '\\' . __NAMESPACE__ . '\\Transforms\\' . str_replace( '_', '', ucwords( $field['type'], '_' ) );
```
Comment (mitogh) [inline src/All.php]: Can we create a method for this or a descriptive variable name since I'm not sure what the line is doing just by reading the line.

File: src/All.php [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+		if ( method_exists( $class, 'apply' ) ) {
```
Comment (mitogh) [inline src/All.php]: I think if we wrap both inside of an `if/else` make it easier to follow on the reading .

File: src/All.php [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+		return $field['value'];
+	}
+
+	/**
+	 * Get all field values.
+	 *
+	 * @param int $target The target object.
+	 * @return array
+	 */
+	private static function get_fields( $target = 0 ) {
```
Comment (mitogh) [inline src/All.php]: - I think this should be called `get_all_fields` instead. 
- What target means?

File: src/All.php [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+		$data = [];
+
+		if ( self::is_active() ) {
+			$fields = get_field_objects( $target );
+
+			if ( $fields ) {
+				foreach ( $fields as $field_name => $field ) {
```
Comment (mitogh) [inline src/All.php]: I think this method is doing two thing the other part of the job is happening inside of this foreach, I think we should split in another method.

File: src/All.php [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+	public static function get_post_fields( $post_id = 0, $include_wp_fields = true ) {
+		if ( $include_wp_fields ) {
+			$wp_fields = [
+				'post_id' => $post_id,
+				'permalink' => get_permalink( $post_id ),
+				'title' => get_the_title( $post_id ),
+				'content' => apply_filters( 'the_content', get_post_field( 'post_content', $post_id ) ),
+			];
+
+			return array_merge( $wp_fields, self::get_fields( $post_id ) );
```
Comment (mitogh) [inline src/All.php]: Why not use `wp_parse_args`?

File: src/All.php [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+		}
+
+		return self::get_fields( $post_id );
```
Comment (mitogh) [inline src/All.php]: Same as Above. I think makes easier to follow if we split this inside of an else clause.

File: src/Transforms/Image.php [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+class Image
+{
+	/**
+	 * Apply the transforms
+	 *
+	 * @param array $field The field.
+	 * @return array
+	 */
+	public static function apply( $field ) {
+		if ( ! ( 'image' === $field['type'] && 'id' === $field['return_format']) ) {
```
Comment (mitogh) [inline src/Transforms/Image.php]: I think if you reverse the comparison operation we can put inside of the if the call to 

```
self::transform_image_fields( $field ); 
```

And just have 1 return instead of 2

File: src/Transforms/Image.php [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+		return $field['value'];
+	}
+
+	/**
+	 * Do the image size transform.
+	 *
+	 * @param string $field			Field
+	 * @return array
+	 */
+	public static function transform_image_fields( &$field ) {
```
Comment (mitogh) [inline src/Transforms/Image.php]: I think we should avoid passing a param by reference since it allow to overwrite the passed function I think the function should return the data as a new array and assign that on the function call in my opinion makes things more easier to understand at a first glance.

File: src/Transforms/Repeater.php [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+	 */
+	public static function transform_to_object( &$field ) {
+		if ( 1 !== count( $field['value'] ) ) {
+			return $field['value'];
+		}
+
+		$as_array = apply_filters( 'ln_acf_repeater_as_array', true, $field );
+
+		$field['value'] = $as_array ? $field['value'] : $field['value'][0];
+	}
```
Comment (mitogh) [inline src/Transforms/Repeater.php]: I think we should return something from there isn't?

File: src/Transforms/Repeater.php [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+
+	/**
+	 * Transform to an object if the filter is set and there is only 1 item.
+	 *
+	 * @param array $field Field.
+	 * @return mixed
+	 */
+	public static function transform_to_object( &$field ) {
+		if ( 1 !== count( $field['value'] ) ) {
```
Comment (mitogh) [inline src/Transforms/Repeater.php]: I think `! empty` is more readable.

File: src/All.php (hunk 1 above)
Comment (adamf321) [inline src/All.php]: We also have Single for a single value:

```
Leean\Acf\All
Leean\Acf\Single
```

File: src/All.php (hunk 2 above)
Comment (adamf321) [inline src/All.php]: Would $class_name be better?

File: src/All.php (hunk 3 above)
Comment (adamf321) [inline src/All.php]: I think this is ok... it's quite normal to have a fallback in case no other return is reached. In this case theres' only on if/else, but if the function were more complex this would make even more sense.

File: src/All.php (hunk 4 above)
Comment (adamf321) [inline src/All.php]: - I think the "All" name of the class implies that all functions get all fields. Don't you?
- The target id - could be a post, a widget, options, etc...

File: src/All.php (hunk 5 above)
Comment (adamf321) [inline src/All.php]: I think things get more complicated if we split them out. Eg how do we handle the if/else within the foreach? Do we pass the $data param by value?

File: src/All.php (hunk 6 above)
Comment (adamf321) [inline src/All.php]: Because wp_parse_args is for merging an array of args with a set of default values. In this case I'm just merging 2 arrays.

File: src/All.php (hunk 7 above)
Comment (adamf321) [inline src/All.php]: Same as above... I think it's quite standard practice to have a fallback return at the bottom of a function.

File: src/Transforms/Image.php (hunk 8 above)
Comment (adamf321) [inline src/Transforms/Image.php]: Yes, much nicer, thanks

File: src/Transforms/Image.php (hunk 9 above)
Comment (adamf321) [inline src/Transforms/Image.php]: My idea here is that it makes it easier to chain the transforms (eg in Repeater we have 2). The transform update the field['value'] param, then it's passed to the next one to do the same. I think it gets a bit more confusing if we return a value then have to pass it to the next transform - for one thing we would have the $value and $field['value'] both with different values.

But I'm open to changing this if you think there's a better way?

File: src/Transforms/Repeater.php (hunk 10 above)
Comment (adamf321) [inline src/Transforms/Repeater.php]: See above

File: src/Transforms/Repeater.php (hunk 11 above)
Comment (adamf321) [inline src/Transforms/Repeater.php]: That doesn't work because I'm looking for exactly 1.  !empty would continue if there were 2 or 3 which I don't want.

File: src/All.php (hunk 2 above)
Comment (mitogh) [inline src/All.php]: I think so but I think a method should be even better since at the moment is a little bit hard to follow the code in that single line.

File: src/All.php (hunk 1 above)
Comment (mitogh) [inline src/All.php]: Both seems methods to me instead of classes at least just by taking a look at the name.

File: src/All.php (hunk 3 above)
Comment (mitogh) [inline src/All.php]: Ok.

File: src/All.php (hunk 2 above)
Comment (mitogh) [inline src/All.php]: Do we have an update on this?

File: src/All.php (hunk 4 above)
Comment (mitogh) [inline src/All.php]: Ok then target should be named `target_id`  instead?

File: src/All.php (hunk 5 above)
Comment (mitogh) [inline src/All.php]: I don't think so since if we do that we are creating a pure function that receives a single item instead, and yes I think is how function should work having and input generates an output.

File: src/All.php (hunk 6 above)
Comment (mitogh) [inline src/All.php]: Right!

File: src/All.php (hunk 7 above)
Comment (mitogh) [inline src/All.php]: Ok!

File: src/Transforms/Repeater.php (hunk 11 above)
Comment (mitogh) [inline src/Transforms/Repeater.php]: mmm on the conditional I see you are looking for an  `n`  that is not 1 and that will work with 2 or 3 the values that you don't want.

File: src/Transforms/Repeater.php [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+ */
+class Repeater
+{
+	/**
+	 * Apply the transforms
+	 *
+	 * @param array $field The field.
+	 * @return array
+	 */
+	public static function apply( $field ) {
```
Comment (mitogh) [inline src/Transforms/Repeater.php]: I think is very similar to the function above, can we use only one function. I think having an abstract class called `Transform`  might help with the task.

File: src/All.php (hunk 1 above)
Comment (adamf321) [inline src/All.php]: The classes are really just groups of methods. So all the methods in the All class retrieve all fields, and the ones in the Single class for single.
However, I just thought we could have one class called Acf and overload the functions (the $field_name would be optional). I think this is better, right?

File: src/All.php (hunk 2 above)
Comment (adamf321) [inline src/All.php]: updated

File: src/All.php (hunk 4 above)
Comment (adamf321) [inline src/All.php]: Ok. updated

File: src/All.php (hunk 5 above)
Comment (adamf321) [inline src/All.php]: Let me refactor it all based on the above

File: src/Transforms/Repeater.php (hunk 11 above)
Comment (adamf321) [inline src/Transforms/Repeater.php]: Yes, and exiting the function early if it's nay of the values i don't want.

File: src/Transforms/Repeater.php (hunk 12 above)
Comment (adamf321) [inline src/Transforms/Repeater.php]: This doesn't make sense afaik, because it's a static method. Static methods can't be implemented in a child class, in fact I don't think you can even declare an abstract static method.

**Comment by adamf321** [comment]
@mitogh pls review now

**Comment by mitogh** [comment]
@adamf321 can we fix travis.

Thanks.

File: src/Transforms/Repeater.php (hunk 12 above)
Comment (mitogh) [inline src/Transforms/Repeater.php]: Yes, that's true. I think you should just have a static API and methods and objects behind or just plain inheritance.
